Peer health checks

shubham3121 opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

- track ping status, ping message and ping timestamp at NodePeer level
- add a utility track peer route healthcheck
- add a network service to find peer
- add an integration test for peers health check

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

shubham3121 committed:
track ping status, ping message and ping timestamp at NodePeer level
- add a utility track peer route healthcheck
- add a network service to find peer

shubham3121 committed:
network::utils: clean up use of network stash

shubham3121 committed:
rename find peer service to check peer association

shubham3121 committed:
dummy commit

shubham3121 committed:
Merge branch 'dev' into shubham/peer-health-checks

shubham3121 committed:
make PeerHealthCheck a class
- run the healthcheck for peer in a thread

khoaguin committed:
remove unnecessary import

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
[tests/network] add an integration test for peers health check

khoaguin committed:
[syft/network] done checking `PEER_ASSOCIATION_PENDING` for `check_peer_association`
- add integration test for `check_peer_association`
- update `protocol_veresion.json` for `NodePeer`

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
[tests/integration] fix `test_peer_health_check`

khoaguin committed:
[syft/network] make a dedicated function to get association requests by peer id
- Add request delete service method
- When deleting a peer, also delete all the association requests from this peer

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
[syft/network] small fixes for `peer_route_healthcheck`

khoaguin committed:
[syft/network] add some error handlings in `peer_route_healthcheck`

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin commented:
TODO in this PR:
- Integrate `background_tasks` in orchestra

2. for `run_peer_health_checks` background task
- What is the ping status of a peer if association request is submitted but not accepted?
- What is the ping status if the association request is accepted?
- If one peer node goes down, what is the ping status (may not be checked with a Python Node. Try starting a server with HTTP connection)
- what is the ping status when a peer is deleted -> create remote client on that node peer -> check peer list -> peer not found

3. Since we can get domain health status now with the new peers health check functionalities, we can just grab the online domains from the database and no longer need to check it like currently in `DomainRegistry.online_domains` using a `ThreadPoolExecutor`

khoaguin committed:
[syft/network] integrate `run_peer_health_checks` as a background task in a `Node`

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

khoaguin committed:
[syft/network] - return `SyftError` if there are pending request from a peer or if it already is a peer

khoaguin committed:
[syft/network] only check the last request to see if a peer's association status is pending

khoaguin commented on packages/syft/src/syft/service/network/network_service.py:
```diff
@@ -237,19 +237,39 @@ def add_peer(
                 message="verify_key does not match the remote node's verify_key for add_peer"
             )
 
+        # check if the peer already is a node peer
+        existed_peer = self.stash.get_by_uid(context.node.verify_key, peer.id)
+        if existed_peer.is_err():
+            return SyftError(
+                message=f"Failed to query peer from stash: {existed_peer.err()}"
+            )
+        if isinstance(existed_peer.ok(), NodePeer):
+            return SyftError(
+                message=f"The peer '{peer.name}' is already associated with '{context.node.name}'"
+            )
+
+        # check if the peer already submitted an association request
+        association_requests: list[Request] = self._get_association_requests_by_peer_id(
+            context=context, peer_id=peer.id
+        )
+        if (
+            association_requests
+            and association_requests[-1].status == RequestStatus.PENDING
+        ):
+            return SyftError(
+                message="There is already a pending association request for this peer"
+            )
+        # only create and submit a new request if there is no requests yet
+        # or all previous requests have been rejected
```
@kiendang please review this

khoaguin replied:
This is the fix for the association requests that we discussed earlier
- Check the last association request by timestamp (`updated_at`) and assign `PEER_ASSOCIATION_PENDING` only when the last request is pending
- Do not allow new association requests if there is a pending request from a peer / peer already associated
- Only create new requests if the last request from the peer has been denied or the peer does not submit any request yet

kiendang replied:
Looks good to me apart from a nitpick change. Needs a second pair of eyes from @shubham3121 too.

kiendang commented on packages/syft/src/syft/service/network/network_service.py:
```diff
@@ -228,19 +237,39 @@ def add_peer(
 
+        # check if the peer already is a node peer
+        existed_peer = self.stash.get_by_uid(context.node.verify_key, peer.id)
+        if existed_peer.is_err():
+            return SyftError(
+                message=f"Failed to query peer from stash: {existed_peer.err()}"
+            )
```
We use python>=3.10 now so can safely do this go-like thing

```suggestion
        if (err := existed_peer.is_err()):
            return SyftError(
                message=f"Failed to query peer from stash: {err}"
            )
```

but I'm not sure about including the error/exception here. @shubham3121?

kiendang committed:
Clean up errors

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Handle repeated association requests

If an association request is made to a peer when another association
request has been previously sent:
- return SyftSuccess if the 2 peers were already connected (the previous
  request was accepted)
- return the Request object if the previous request is still pending

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Update peer information if changed

in add_peer

khoaguin committed:
fix linting

khoaguin committed:
[syft/network] make `PeerHealthCheckTask` serializable

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

kiendang committed:
Close background thread on node shutdown

khoaguin committed:
[test/integration] update `test_peer_health_check`

shubham3121 commented on packages/hagrid/hagrid/orchestra.py:
```diff
@@ -272,6 +272,7 @@ def deploy_to_python(
         "n_consumers": n_consumers,
         "create_producer": create_producer,
         "association_request_auto_approval": association_request_auto_approval,
+        "background_tasks": True,
```
suggestion: we need to propagate this `background_tasks` flag to orchestra.launch

shubham3121 replied:
@khoaguin I think we missed addressing this comment. Need to integrate background_tasks to def launch and then propogate it to other deployment types.

khoaguin replied:
Thank you very much @shubham3121
Integrating `background_tasks` flag to `orchestra.launch` in this commit https://github.com/OpenMined/PySyft/pull/8795/commits/8182a1480a6dfcf3f56568428b35aa644e475807

shubham3121 commented on packages/syft/src/syft/service/network/network_service.py:
```diff
@@ -170,6 +179,33 @@ def exchange_credentials_with(
+            existing_peer_result.is_ok()
+            and (existing_peer := existing_peer_result.ok()) is not None
+        ):
+            msg = [f"Routes already exchanged."]
```
nitpick: remove fstring `f`

shubham3121 commented on packages/syft/src/syft/service/network/node_peer.py:
```diff
@@ -66,6 +75,9 @@ class NodePeer(SyftObject):
     admin_email: str
+    ping_status: NodePeerConnectionStatus | None = None
+    ping_status_message: str | None = None
+    pinged_timestamp: DateTime | None = None
```
Can we add these three vars to __repr__attrs__

shubham3121 commented on packages/syft/src/syft/service/network/utils.py:
```diff
@@ -0,0 +1,123 @@
+            )
+
+            if result.is_err():
+                logging.info(f"Failed to update peer in stash: {result.err()}")
```
can we replace use of logging with loguru

shubham3121 replied:
```python
from loguru import logger
logger.info()
```

shubham3121 commented on packages/syft/src/syft/service/request/request_service.py:
```diff
@@ -288,6 +288,19 @@ def save(
+        path="request.delete_by_id",
+        name="delete_by_id",
+    )
+    def delete_by_id(
```
nitpick: rename to `delete_by_uid`

khoaguin committed:
[syft/network] stop serializing `PeerHealthCheckTask`'s thread

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

shubham3121 commented:
Great work @khoaguin and @kiendang !!! Left a few minor comments.

khoaguin committed:
[syft/network] small fixes

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

khoaguin committed:
[test/unit] mark `test_in_memory_action_graph_store_init` as flaky

khoaguin committed:
[test/network] fix gateway local test since now we return SyftSuccess if a node is already a peer instead of a Request object

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
[syft/network] stop updating a route's priority if it already exists

khoaguin committed:
[test/integration] fix `test_add_route`. Also test `update_route_priority` service

khoaguin committed:
[test/integration] fix `test_add_route_on_peer`

khoaguin committed:
[syft/network] Also check remotely if the self node already exists as a peer for `exchange_credentials_with` service

khoaguin committed:
[refactor] reformatting some code for easier reading

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
[test/integration] `gateway_local_test` now passes locally

khoaguin committed:
[test/integration] add a check for peer connection status in gateway tests

khoaguin committed:
[syft/network] - improve returned success message for `EnclaveClient.connect_to_gateway()`
- add error handling for `Node.settings()`

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
[notebook/tutorial] remove repeated code in `Enclave-single-notebook-high-low-network` notebook

khoaguin committed:
[syft/network] fix linting

khoaguin committed:
[syft/network] increase `repeat_time` for `PeerHealthCheckTask` to 60 secs which seem to fix the SQLiteBackingStore error `Cannot operate on a closed database"

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
Merge branch 'dev' into shubham/peer-health-checks

khoaguin committed:
[syft/network] - integrate peers health check into orchestra and for webserver nodes
- `background_tasks` for a Node defaults to `False`
- add some tests to `gateway_local_test`

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

khoaguin committed:
[test/integration] test peer connection status for k8s nodes

shubham3121 reviewed (commented):
Great work @khoaguin and thanks for the help @kiendang !! 🎉 💯
